### ISSUE TITLE
Tasks stay queued forever

### ISSUE THREAD

**Issue by mvach** [cloudfoundry/bosh#2429]
**Describe the bug**
We managed to queue tasks on the director that are afterwards ignored and never processed. We got into the situation during the initial rollout of our database credential rotation where we removed the login privilege from the previously used postgres role.

Our bosh deployment process consist of multiple steps and the credential rotation happens very early in the process where as the real deployment of bosh happens relatively late. In the time during these two steps some task have been created and seem to be queued forever now.

Now we are wondering why bosh isn't able to process such tasks.
Here the relevant task from the tasks table:
```
boshdb=> select id, state, timestamp, result, checkpoint_time, type, deployment_name, started_at, event_output, result_output, context_id from tasks where state = 'queued';
   id    | state  |         timestamp          | result |      checkpoint_time       |           type           | deployment_name | started_at | event_output | result_output | context_id
---------+--------+----------------------------+--------+----------------------------+--------------------------+-----------------+------------+--------------+---------------+------------
 7216738 | queued | 2023-01-18 13:20:41.166006 |        | 2023-01-18 13:20:41.166013 | vms                      | haproxy         |            |              |               |
 7216737 | queued | 2023-01-18 13:18:59.553377 |        | 2023-01-18 13:18:59.553385 | vms                      | cf              |            |              |               |
 7216736 | queued | 2023-01-18 13:10:36.593763 |        | 2023-01-18 13:10:36.593774 | vms                      | haproxy         |            |              |               |
 7216735 | queued | 2023-01-18 13:09:41.234805 |        | 2023-01-18 13:09:41.234814 | vms                      | cf              |            |              |               |
 7216734 | queued | 2023-01-18 13:07:30.192177 |        | 2023-01-18 13:07:30.192185 | vms                      | cf              |            |              |               |
 7216733 | queued | 2023-01-18 13:04:30.887932 |        | 2023-01-18 13:04:30.887941 | vms                      | bosh-sf         |            |              |               |
 7216732 | queued | 2023-01-18 13:00:00.206581 |        | 2023-01-18 13:00:00.206585 | scheduled_events_cleanup |                 |            |              |               |
 7216731 | queued | 2023-01-18 12:59:57.391192 |        | 2023-01-18 12:59:57.3912   | vms                      | haproxy         |            |              |               |
 7216730 | queued | 2023-01-18 12:55:59.062308 |        | 2023-01-18 12:55:59.062316 | vms                      | cf              |            |              |               |
 7216729 | queued | 2023-01-18 12:50:51.388166 |        | 2023-01-18 12:50:51.388177 | vms                      | cf              |            |              |               |
 7216728 | queued | 2023-01-18 12:49:55.404581 |        | 2023-01-18 12:49:55.40459  | vms                      | haproxy         |            |              |               |
 7216727 | queued | 2023-01-18 12:39:54.033932 |        | 2023-01-18 12:39:54.033939 | vms                      | haproxy         |            |              |               |
 7216726 | queued | 2023-01-18 12:35:41.388784 |        | 2023-01-18 12:35:41.388792 | vms                      | cf              |            |              |               |
 7216725 | queued | 2023-01-18 12:34:23.450374 |        | 2023-01-18 12:34:23.450382 | vms                      | bosh-sf         |            |              |               |
 7216724 | queued | 2023-01-18 12:33:31.689413 |        | 2023-01-18 12:33:31.689421 | ssh                      | cf              |            |              |               |
 7216723 | queued | 2023-01-18 12:33:31.461821 |        | 2023-01-18 12:33:31.46183  | ssh                      | cf              |            |              |               |
(16 rows)  
```

We are currently cleaning up these tasks on multiple landscapes and noticed that the timestamp, and checkpoint_time is not available via the bosh task command. Which means that we have to query the bosh database to figure out what queued task is existing for a long while and can be safely canceled. May be that information can be added to the tasks command https://github.com/cloudfoundry/bosh/issues/2431. 

**Expected behavior**
Bosh should cleanup queued task after a while

**Comment by beyhan** [comment]
Hi @mvach,
the pr https://github.com/cloudfoundry/bosh/pull/2413 has explanation when tasks can stay forever in the `tasks` table. Can you look into the pr and check the situation on your side. Interesting here is what does the `delayed job table` shows for the listed tasks above.

**Comment by mvach** [comment]
Hi,
it seems that pr https://github.com/cloudfoundry/bosh/pull/2413 is not working for us.
I checked another landscape which is using bosh v277.2.0 (which is containing https://github.com/cloudfoundry/bosh/pull/2413)

We found:
```
bosh=> select count(*) from tasks where state = 'queued';
 count
-------
   456
(1 row)

bosh=> select count(*) from delayed_jobs;
 count
-------
     1
```

Here some details from the tasks table:
```
bosh=> select id, state, timestamp, result, checkpoint_time, type, started_at, event_output, result_output from tasks where state = 'queued';
   id    | state  |         timestamp          | result |      checkpoint_time       |              type               | started_at | event_output | result_output
---------+--------+----------------------------+--------+----------------------------+---------------------------------+------------+--------------+---------------
 1772215 | queued | 2022-11-30 07:00:42.294603 |        | 2022-11-30 07:00:42.298529 | snapshot_deployment             |            |              |
 1770520 | queued | 2022-11-28 07:00:35.647603 |        | 2022-11-28 07:00:35.649928 | snapshot_deployment             |            |              |
 1766504 | queued | 2022-11-23 07:00:31.546199 |        | 2022-11-23 07:00:31.615641 | snapshot_deployment             |            |              |
 1764382 | queued | 2022-11-21 07:00:42.525964 |        | 2022-11-21 07:00:42.562274 | snapshot_deployment             |            |              |
 1762927 | queued | 2022-11-18 07:00:14.844616 |        | 2022-11-18 07:00:14.855009 | snapshot_deployment             |            |              |
 1756861 | queued | 2022-11-13 07:00:46.253871 |        | 2022-11-13 07:00:46.271164 | snapshot_deployment             |            |              |
 1754992 | queued | 2022-11-10 07:00:28.243157 |        | 2022-11-10 07:00:28.244836 | snapshot_deployment             |            |              |
 1753925 | queued | 2022-11-09 07:00:30.620977 |        | 2022-11-09 07:00:30.64799  | snapshot_deployment             |            |              |
 1751904 | queued | 2022-11-07 07:01:05.027346 |        | 2022-11-07 07:01:05.397311 | snapshot_deployment             |            |              |
 1746550 | queued | 2022-10-31 12:42:31.894807 |        | 2022-10-31 12:42:31.89482  | vms                             |            |              |
 1746399 | queued | 2022-10-31 12:32:47.722951 |        | 2022-10-31 12:32:47.722979 | vms                             |            |              |
 1746373 | queued | 2022-10-31 12:31:55.733429 |        | 2022-10-31 12:31:55.73345  | vms                             |            |              |
 1745972 | queued | 2022-10-31 07:00:42.655175 |        | 2022-10-31 07:00:42.659767 | snapshot_deployment             |            |              |
 1745948 | queued | 2022-10-31 07:00:01.03437  |        | 2022-10-31 07:00:01.0357   | snapshot_deployment             |            |              |
 1745909 | queued | 2022-10-30 07:00:40.399948 |        | 2022-10-30 07:00:40.443484 | snapshot_deployment             |            |              |
 1744597 | queued | 2022-10-28 07:00:45.566976 |        | 2022-10-28 07:00:45.568774 | snapshot_deployment             |            |              |
 1744591 | queued | 2022-10-28 07:00:39.261244 |        | 2022-10-28 07:00:39.263472 | snapshot_deployment             |            |              |
 1744590 | queued | 2022-10-28 07:00:32.285492 |        | 2022-10-28 07:00:36.837826 | snapshot_deployment             |            |              |
 1744589 | queued | 2022-10-28 07:00:30.980169 |        | 2022-10-28 07:00:31.063769 | snapshot_deployment             |            |              |
 1743464 | queued | 2022-10-27 07:00:09.917517 |        | 2022-10-27 07:00:09.918874 | snapshot_deployment             |            |              |
 1742591 | queued | 2022-10-26 12:40:56.435684 |        | 2022-10-26 12:40:56.437433 | vms                             |            |              |
```

**Comment by beyhan** [comment]
Hi @mvach,
Do you mean that the [task cleanup](https://github.com/cloudfoundry/bosh/blob/8ccda24c54029c97b2ca580be47c9fdf77df226a/src/bosh-director/lib/bosh/director/jobs/scheduled_tasks_cleanup.rb) which was implemented for that pr doesn't work for you? 

Were the tasks executed at all? In https://github.com/cloudfoundry/bosh/pull/2413 the tasks has been executed and the updated in the `tasks` table failed. I'm not sure what is the state in this case, but I was expected at least that the [task cleanup](https://github.com/cloudfoundry/bosh/blob/8ccda24c54029c97b2ca580be47c9fdf77df226a/src/bosh-director/lib/bosh/director/jobs/scheduled_tasks_cleanup.rb] will clean the tasks.

**Comment by mvach** [comment]
Hi @beyhan,
exactly that's the case the task cleanup ist not working for us.

**Comment by beyhan** [comment]
do you know why it didn't clean the tasks? Is there an issue with the cleanup?

**Comment by rkoster** [comment]
Have you observed tasks getting in this state after rolling out: https://github.com/cloudfoundry/bosh/pull/2413?

**Comment by rkoster** [comment]
Closing due to inactivity, feel free to reopen when needed.

**Comment by beyhan** [comment]
Seems to be still relevant.

**Comment by mvach** [comment]
We noticed that these hanging jobs are queued on 8ca2e397-6205-405b-b7c1-4a292263e271.
On the other hand the ENV of our worker processes contain the variable `QUEUE=normal,urgent`

Could that be the issue that the queues of such jobs are not existing?

**Comment by nouseforaname** [comment]
From memory I'd assume that the '8ca2e397-6205-405b-b7c1-4a292263e271' is the Bosh worker thread that picked up the task. I do not know if that correlates to anything in Bosh or of that is a delayed job worker id. If those tasks are marked as queued in delayed job by id XYZ but that worker doesn't exist, that would explain why they're stuck?

**Comment by mvach** [comment]
Hmmm, it might be hard to figure out why those workers died. Sounds like this requires a schedule_job_cleanup.rb to delete orphaned jobs. That would simply drop orphaned jobs from delayed_jobs table, which then in turn is set to error state by the scheduled_tasks_cleanup.rb.

Would you agree?

**Comment by mvach** [comment]
Hi,
as discussed a while a we tried to reproduce the issue.

We triggered deployments on a bosh with stopped workers. As soon as we started a worker afterwards bosh processed the related task.
We triggered a deployment and stopped the executing worker while it was processing. Such tasks failed with timeout.

Additionally we noticed that
1. all these tasks had a started_at timestamp as soon as they started the processing.
2. all these tasks where running on the queue 'normal'

It is still unclear how bosh is assigning tasks from time to time to a not existing queue which has a uuid as name.
